src/cmd/s3checker/pkg/s3/s3.go

```diff
@@ -0,0 +1,210 @@
+		Secure: !opt.DoNotUseTLS,
+		Region: opt.Region,
+	}
+
+	minioOpts.Transport = getCustomTransport(opt)
+
+	cli, err := minio.New(opt.Endpoint, minioOpts)
+	if err != nil {
+		return nil, clues.Wrap(err, "creating client")
+	}
+
+	return &Client{
+		client:     cli,
+		bucketName: opt.BucketName,
+		prefix:     opt.Prefix,
+	}, nil
+}
+
+func maybeAddObj(prefix string, obj ObjInfo, m map[string]ObjInfo) {
+	// We've already found an item to check.
+	if _, ok := m[prefix]; ok {
+		return
+	}
+
+	m[prefix] = obj
+}
+
+func (c *Client) ListUntilAllFound(
+	ctx context.Context,
+	wantedPrefixes []string,
+	alsoFindDeleted bool,
+) (map[string]ObjInfo, map[string]ObjInfo, error) {
+	notDeleted := map[string]ObjInfo{}
+	deleted := map[string]ObjInfo{}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	objChan := c.client.ListObjects(
+		ctx,
+		c.bucketName,
+		minio.ListObjectsOptions{
+			Prefix:       c.prefix,
+			WithVersions: alsoFindDeleted,
+		})
+
+	for obj := range objChan {
+		if err := obj.Err; err != nil {
+			return notDeleted, deleted, clues.Wrap(err, "searching object list")
+		}
+
+		// Getting retention info for a deletion marker isn't allowed.
+		if obj.IsDeleteMarker {
+			continue
+		}
+
+		var (
+			matchesPrefix string
+			trimmedKey    = obj.Key[len(c.prefix):]
+		)
+
+		for _, prefix := range wantedPrefixes {
+			if strings.HasPrefix(trimmedKey, prefix) {
+				matchesPrefix = prefix
+				break
+			}
+		}
+
+		// Item doesn't have a prefix that matches anything we want.
+		if len(matchesPrefix) == 0 {
+			continue
+		}
+
+		objI := ObjInfo{
+			Key:     obj.Key,
+			Version: obj.VersionID,
+		}
+
+		// Since we skip deletion markers above, if the item is the latest version
+		// then it's considered not deleted. If it's not the latest then it's likely
+		// deleted. Not asking for versions always sets IsLatest to false.
+		if !alsoFindDeleted || obj.IsLatest {
+			maybeAddObj(matchesPrefix, objI, notDeleted)
+		} else {
+			maybeAddObj(matchesPrefix, objI, deleted)
+		}
+
+		// Break if we've found all the non-deleted items we're looking for and
+		// either we're not looking for deleted items or we've found all the deleted
+		// items we're looking for.
+		if len(notDeleted) == len(wantedPrefixes) &&
+			(!alsoFindDeleted || len(deleted) == len(wantedPrefixes)) {
+			break
+		}
+	}
+
+	return notDeleted, deleted, nil
+}
+
+func (c *Client) ObjectRetention(
+	ctx context.Context,
+	obj ObjInfo,
+) (*minio.RetentionMode, *time.Time, error) {
+	mode, retainUntil, err := c.client.GetObjectRetention(
+		ctx,
+		c.bucketName,
+		obj.Key,
+		obj.Version)
+	if err != nil {
+		// Unfortunately they don't have a sentinel type that we can compare
+		// against. We can check the error code though.
+		var e minio.ErrorResponse
+
+		if errors.As(err, &e) {
+			if e.Code == noLockCode {
+				return nil, nil, clues.Stack(ErrNoRetention, err)
+			}
+		}
+	}
+
+	return mode, retainUntil, clues.Wrap(err, fmt.Sprintf(
```

Comment:
Not sure if we should do `clues.Wrap(...).With(..., ...)` here and expand at the other side.